RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu

```diff
@@ -633,79 +633,65 @@ namespace pixelgpudetails {
-
-    // std::cout << "found " << nModulesActive << " Modules active" << std::endl;
-
-    // In order to avoid the cudaStreamSynchronize, create a new kernel which launches countModules and findClus.
-    cudaStreamSynchronize(stream.id());
-    
-    threadsPerBlock = 256;
-    blocks = nModulesActive;
-
-    /*
-    std::cout
-      << "CUDA findClus kernel launch with " << blocks
-      << " blocks of " << threadsPerBlock << " threads\n";
-    */
-
-    cudaCheck(cudaMemsetAsync(clusInModule_d, 0, (MaxNumModules)*sizeof(uint32_t), stream.id()));
-
-    findClus<<<blocks, threadsPerBlock, 0, stream.id()>>>(
-                 moduleInd_d,
-                 xx_d, yy_d, adc_d,
-                 moduleStart_d,
-                 clusInModule_d, moduleId_d,
-                 clus_d,
-                 wordCounter
-    );
-
-    // clusters
-    cudaCheck(cudaMemcpyAsync(clus_h, clus_d, wordCounter*sizeof(uint32_t), cudaMemcpyDefault, stream.id()));
-
-    cudaStreamSynchronize(stream.id());
-    cudaCheck(cudaGetLastError());
-
-   } // end clusterizer scope
+    {
+      // clusterizer ...
+      using namespace gpuClustering;
+      int threadsPerBlock = 256;
+      int blocks = (wordCounter + threadsPerBlock - 1) / threadsPerBlock;
+
+
+      gpuCalibPixel::calibDigis<<<blocks, threadsPerBlock, 0, stream.id()>>>(
+          moduleInd_d,
+          xx_d, yy_d, adc_d,
+          gains,
+          wordCounter);
+      cudaCheck(cudaGetLastError());
+
+      // calibrated adc
+      cudaCheck(cudaMemcpyAsync(adc_h, adc_d, wordCounter*sizeof(uint16_t), cudaMemcpyDefault, stream.id()));
+
+      /*
+         std::cout
+         << "CUDA countModules kernel launch with " << blocks
+         << " blocks of " << threadsPerBlock << " threads\n";
+       */
+
+      cudaCheck(cudaMemsetAsync(moduleStart_d, 0x00, sizeof(uint32_t), stream.id()));
+
+      countModules<<<blocks, threadsPerBlock, 0, stream.id()>>>(moduleInd_d, moduleStart_d, clus_d, wordCounter);
+      cudaCheck(cudaGetLastError());
+
+      // read the number of modules into a data memeber, used by getProduct())
+      cudaCheck(cudaMemcpyAsync(&nModulesActive, moduleStart_d, sizeof(uint32_t), cudaMemcpyDefault, stream.id()));
+
+      threadsPerBlock = 256;
+      blocks = MaxNumModules;
+      /*
+         std::cout << "CUDA findClus kernel launch with " << blocks
+                   << " blocks of " << threadsPerBlock << " threads\n";
+       */
+      cudaCheck(cudaMemsetAsync(clusInModule_d, 0, (MaxNumModules)*sizeof(uint32_t), stream.id()));
+      findClus<<<blocks, threadsPerBlock, 0, stream.id()>>>(
+          moduleInd_d,
+          xx_d, yy_d,
+          moduleStart_d,
+          clusInModule_d, moduleId_d,
+          clus_d,
+          wordCounter);
+      cudaCheck(cudaGetLastError());
+
+      // clusters
+      cudaCheck(cudaMemcpyAsync(clus_h, clus_d, wordCounter*sizeof(uint32_t), cudaMemcpyDefault, stream.id()));
+      cudaCheck(cudaGetLastError());
```

Comment:
I believe the last `cudaGetLastError()` is not needed after removing the `cudaStreamSynchronize()`, possible errors in queueing the async copy should get caught already by its return value.

Reply:
Fixed, thanks